File: sport-ecommerce-be/src/main/java/com/sport/ecommerce/modules/coupon/service/CouponService.java
package com.sport.ecommerce.modules.coupon.service;

import com.sport.ecommerce.modules.coupon.dto.request.ApplyCouponRequest;
import com.sport.ecommerce.modules.coupon.dto.request.CreateCouponRequest;
import com.sport.ecommerce.modules.coupon.dto.request.UpdateCouponRequest;
import com.sport.ecommerce.modules.coupon.dto.response.ApplyCouponResponse;
import com.sport.ecommerce.modules.coupon.dto.response.CouponResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface CouponService {

    Page<CouponResponse> getCoupons(Pageable pageable);

    CouponResponse getCouponById(Long id);

    CouponResponse getCouponByCode(String code);

    CouponResponse createCoupon(CreateCouponRequest request);

    CouponResponse updateCoupon(Long id, UpdateCouponRequest request);

    void deleteCoupon(Long id);

    ApplyCouponResponse applyCoupon(ApplyCouponRequest request);
}

File: sport-ecommerce-fe/src/app/features/cart/services/coupon.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { BaseHttpService } from '../../../core/services/base-http.service';
import { COUPON_API } from '../../../core/constants/api-path.constant';
import { ApiResponse } from '../../../models/api-response.model';
import {
  ApplyCouponRequest,
  ApplyCouponResponse,
  CouponResponse,
} from '../../../models/coupon.model';

@Injectable({ providedIn: 'root' })
export class CouponService {
  private readonly http = inject(BaseHttpService);

  applyCoupon(request: ApplyCouponRequest): Observable<ApiResponse<ApplyCouponResponse>> {
    return this.http.post<ApplyCouponResponse>(COUPON_API.APPLY, request);
  }

  getCouponByCode(code: string): Observable<ApiResponse<CouponResponse>> {
    return this.http.get<CouponResponse>(COUPON_API.BY_CODE(code));
  }
}
